ddtrace/contrib/jinja2/patch.py

```diff
@@ -0,0 +1,93 @@
+import jinja2
+from wrapt import wrap_function_wrapper as _w
+
+from ddtrace import config
+
+from ...ext import http
+from ...utils.formats import get_env
+from ...pin import Pin
+from ...utils.wrappers import unwrap as _u
+from .constants import DEFAULT_TEMPLATE_NAME
+
+
+# default settings
+config._add('jinja2', {
+    'service_name': get_env('jinja2', 'service_name', None),
+})
+
+
+def patch():
+    if getattr(jinja2, '__datadog_patch', False):
+        # already patched
+        return
+    setattr(jinja2, '__datadog_patch', True)
+    Pin(
+        service=config.jinja2['service_name'],
+        _config=config.jinja2,
```

Comment:
This just broke our deploy. `config.jinja2['service_name']` is `None` by default, `app` and `app_type` are not set which sends something like this (recorded with tcpdump and decoded from msgpack) to `dd-agent`:

```
`{None: {'app': None, 'app_type': None},
 'aws': {'app': 'aws', 'app_type': 'web'},
 'flask': {'app': 'flask', 'app_type': 'web'}}`
```

The response from `dd-agent` is `400 Bad Request`: `msgp: attempted to decode type "nil" with method for "str"`

We cannot upgrade `dd-agent` right now so we cannot use the latest version of `ddtrace` (`0.16.0`).

Reply:
Sorry to hear @bancek! Can you elaborate what you mean by breaking your deploy? A 400 from the agent should not cause your app to crash (it should just log).

Reply:
The app (flask) logged `400 Bad Request`, served one request and than stopped processing any further requests. When we disabled `ddtrace` it worked normally. Now we've downgraded `ddtrace` to `0.15.0` and it works again.

Reply:
Strange. The endpoint for services is no longer required and should not cause issues (even if it 400s) and we're in the process of removing the logic from ddtrace (#717).

Can you open an issue providing the relevant logs from the agent and flask along with your python version and relevant `pip freeze` version information so that we can investigate?

In the meantime, you could try [configuring](http://pypi.datadoghq.com/trace/docs/other_integrations.html#jinja2) the jinja service name? However, I have a feeling this won't really work since the `app` and `app_type` are `None` still.

Thanks for reporting!

Reply:
I've tried to set Jinja2 service name but as you mentioned `app` and `app_type` are still `None`.

```
import ddtrace.contrib.jinja2.patch  # import patch so default config (empty) is added for Jinja2
from ddtrace import config
config.jinja2['service_name'] = 'jinja2-templates'
```

Python version (Docker `python:3.7.1-alpine3.8`):

```
Python 3.7.1 (default, Nov 16 2018, 06:29:01)
```

Our `requirements.txt`:

```
certifi==2018.10.15       # via sentry-sdk
ddtrace==0.16.0
flask==1.0.2
itsdangerous==1.1.0
jinja2==2.10
msgpack-python==0.5.6     # via ddtrace
mysqlclient==1.3.13
pycparser==2.19           # via cffi
python-dateutil==2.6.1
sentry-sdk==0.5.5
six==1.11.0
sqlalchemy==1.2.14
urllib3==1.24.1           # via sentry-sdk
werkzeug==0.14.1
wrapt==1.10.11            # via ddtrace
```

TCP traffic recording:

```
00000000  50 55 54 20 2f 76 30 2e  33 2f 73 65 72 76 69 63   PUT /v0. 3/servic
00000010  65 73 20 48 54 54 50 2f  31 2e 31 0d 0a 48 6f 73   es HTTP/ 1.1..Hos
00000020  74 3a 20 64 64 2d 61 67  65 6e 74 3a 38 31 32 36   t: dd-ag ent:8126
00000030  0d 0a 41 63 63 65 70 74  2d 45 6e 63 6f 64 69 6e   ..Accept -Encodin
00000040  67 3a 20 69 64 65 6e 74  69 74 79 0d 0a 43 6f 6e   g: ident ity..Con
00000050  74 65 6e 74 2d 4c 65 6e  67 74 68 3a 20 31 31 37   tent-Len gth: 117
00000060  0d 0a 43 6f 6e 74 65 6e  74 2d 54 79 70 65 3a 20   ..Conten t-Type: 
00000070  61 70 70 6c 69 63 61 74  69 6f 6e 2f 6d 73 67 70   applicat ion/msgp
00000080  61 63 6b 0d 0a 44 61 74  61 64 6f 67 2d 4d 65 74   ack..Dat adog-Met
00000090  61 2d 4c 61 6e 67 3a 20  70 79 74 68 6f 6e 0d 0a   a-Lang:  python..
000000A0  44 61 74 61 64 6f 67 2d  4d 65 74 61 2d 4c 61 6e   Datadog- Meta-Lan
000000B0  67 2d 56 65 72 73 69 6f  6e 3a 20 33 2e 37 2e 31   g-Versio n: 3.7.1
000000C0  0d 0a 44 61 74 61 64 6f  67 2d 4d 65 74 61 2d 4c   ..Datado g-Meta-L
000000D0  61 6e 67 2d 49 6e 74 65  72 70 72 65 74 65 72 3a   ang-Inte rpreter:
000000E0  20 43 50 79 74 68 6f 6e  0d 0a 44 61 74 61 64 6f    CPython ..Datado
000000F0  67 2d 4d 65 74 61 2d 54  72 61 63 65 72 2d 56 65   g-Meta-T racer-Ve
00000100  72 73 69 6f 6e 3a 20 30  2e 31 36 2e 30 0d 0a 0d   rsion: 0 .16.0...
00000110  0a                                                 .
00000111  84 a3 61 77 73 82 a3 61  70 70 a3 61 77 73 a8 61   ..aws..a pp.aws.a
00000121  70 70 5f 74 79 70 65 a3  77 65 62 c0 82 a3 61 70   pp_type. web...ap
00000131  70 c0 a8 61 70 70 5f 74  79 70 65 c0 a5 66 6c 61   p..app_t ype..fla
00000141  73 6b 82 a3 61 70 70 a5  66 6c 61 73 6b a8 61 70   sk..app. flask.ap
00000151  70 5f 74 79 70 65 a3 REDACTED                      p_type.REDACTED
00000161  REDACTED                                82 a3 61   REDACTED      ..a
00000171  70 70 a5 66 6c 61 73 6b  a8 61 70 70 5f 74 79 70   pp.flask .app_typ
00000181  65 a3 77 65 62                                     e.web

00000000  48 54 54 50 2f 31 2e 31  20 34 30 30 20 42 61 64   HTTP/1.1  400 Bad
00000010  20 52 65 71 75 65 73 74  0d 0a 43 6f 6e 74 65 6e    Request ..Conten
00000020  74 2d 54 79 70 65 3a 20  74 65 78 74 2f 70 6c 61   t-Type:  text/pla
00000030  69 6e 3b 20 63 68 61 72  73 65 74 3d 75 74 66 2d   in; char set=utf-
00000040  38 0d 0a 58 2d 43 6f 6e  74 65 6e 74 2d 54 79 70   8..X-Con tent-Typ
00000050  65 2d 4f 70 74 69 6f 6e  73 3a 20 6e 6f 73 6e 69   e-Option s: nosni
00000060  66 66 0d 0a 44 61 74 65  3a 20 4d 6f 6e 2c 20 31   ff..Date : Mon, 1
00000070  39 20 4e 6f 76 20 32 30  31 38 20 31 34 3a 32 38   9 Nov 20 18 14:28
00000080  3a 30 32 20 47 4d 54 0d  0a 43 6f 6e 74 65 6e 74   :02 GMT. .Content
00000090  2d 4c 65 6e 67 74 68 3a  20 35 39 0d 0a 0d 0a      -Length:  59....
0000009F  6d 73 67 70 3a 20 61 74  74 65 6d 70 74 65 64 20   msgp: at tempted 
000000AF  74 6f 20 64 65 63 6f 64  65 20 74 79 70 65 20 22   to decod e type "
000000BF  6e 69 6c 22 20 77 69 74  68 20 6d 65 74 68 6f 64   nil" wit h method
000000CF  20 66 6f 72 20 22 73 74  72 22 0a                   for "st r".
```

I cannot provide you logs from `dd-agent` as they contain sensitive info.

Reply:
No problem about the agent logs, they shouldn't be very relevant besides the 400 anyway.

Thanks for all the info! If you'd like, we can debug quicker if you don't mind joining our public slack here: http://chat.datadoghq.com/

Since you're using flask and our new flask integration landed with the 0.16 release as well, we can try investigating whether the bug is with the jinja integration or the flask.

Can you try running your app with jinja2 disabled? You can do this with `DATADOG_PATCH_MODULES=jinja2:false` or with `patch_all(jinja2=false)`.